bin/DataDownloadAndMerge.pm

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/perl -w
+use strict;
+use warnings;
+
+package DataDownloadAndMerge;
+use FindBin qw/$RealBin/;
+use lib "$RealBin/bin";
+use Utility;
+use File::Basename;
+
+sub new {
+    my $class = shift;
+    my $self  = {};
+    bless $self, $class;
+    return $self;
+}
+
+sub start {
+    my ( $self, $args_dict ) = @_;
+
+    print STDERR "\n\n";
+    print STDERR
+"======================================================================\n";
+    print STDERR
+"    Step4+5: Merge/Downloading from HDFS file system to local directory \n";
+    print STDERR
+"======================================================================\n";
+
+    ################### Time start #########################
+    my $start_time       = time();
+    my $time_stamp_start = Utility::get_time_stamp();
+    print STDERR "\nStarted at:  $time_stamp_start\n";
+
+    ### Paired-end data download
+    $self->paired_end_data($args_dict);
+
+    ### Single-end data download
+    $self->single_end_data($args_dict);
+
+    ################### Time end #########################
+    my $end_time       = time();
+    my $execution_time = Utility::get_execution_time( $start_time, $end_time );
+    my $time_stamp_end = Utility::get_time_stamp();
+    print STDERR "Finished at: $time_stamp_end\n";
+    print STDERR "Duration: $execution_time\n\n";
+}
+
+sub paired_end_data {
+    my ( $self, $args_dict ) = @_;
+
+    $args_dict->{"read_folder"}       = "fastq_paired_end";
+    $args_dict->{"read_type"}         = "pe";
+    $args_dict->{"final_output_file"} = "DistMap_output_Paired_end_reads";
+
+    my $hadoop_exe    = $args_dict->{'hadoop_exe'};
+    my $hdfs_exe      = $args_dict->{"hdfs_exe"};
+    my $readtools     = $args_dict->{"readtools"};
+    my $output        = "/$args_dict->{'random_id'}" . "_output";
+    my $output_folder = "$output/$args_dict->{'read_folder'}" . "_mapping";
+
+    my $file_count = $self->get_file_list(
+"$args_dict->{'output_directory'}/$args_dict->{'random_id'}/$args_dict->{'read_folder'}"
+    );
+
+    my $hdfs_input_folder =
+      "/$args_dict->{'random_id'}" . "_input/fastq_paired_end";
+
+    my @cmd = ();
+    if ( ( system("$hdfs_exe dfs -test -d $hdfs_input_folder") == 0 ) ) {
+        @cmd = `$args_dict->{'hdfs_exe'} dfs -ls $hdfs_input_folder`;
+    }
+
+    my $hdfs_file_count = 0;
+
+    if ( scalar(@cmd) > 1 ) {
+        $hdfs_file_count = $cmd[0];
+        $hdfs_file_count =~ s/\D//g;
+    }
+
+    if ( $file_count > 0 or $hdfs_file_count > 0 ) {
```

Comment:
This is basically DataDownload.pm, with the download command replaced and the target file name taken from DataMerge.pm.  I left the rest intact because I do not yet fully understand it and do not want to break anything.

Reply:
Ok, let's change it in a different PR